Enable pylint for our tests

emlove opened the pull request:
## Description:
This PR enables pylint for our tests directory. A good portion of our tests currently don't pass pylint, so for now they've been blacklisted so we can turn on pylint for the rest of them.

The failing tests were modified with the following bash script:
```bash
sed -i '0,/"""$/s//"""\n# pylint: skip-file/' $(pylint --msg-template='{path}' tests 2>/dev/null | grep '^tests' | uniq)
```

### Notes
We should also be braced for incoming PRs introducing lint failures unless they're rebased/merged on dev.

emlove committed:
Run pylint on our tests directory

OttoWinter commented:
Please see this comment: https://github.com/home-assistant/home-assistant/pull/13093#discussion_r173664662. There are quite a few `ignore`s that should be set up first in order to not break the tests process. I would say we should first try fixing at least *a few* files to see how well the linting works.

Also, we need to think about speed. Right now, we skip most test files. But if we get more test files "pylinted" and remove the skips, the pylint travis job will be incredibly slow (about 2x)

armills commented:
The scope of this PR is only to turn on pylint, especially for new tests coming in. Existing tests that aren't passing need to be dealt with separately in their own PRs.

OttoWinter commented:
Sure, but what if somebody creates a PR with tests after this PR gets merged? They will get *tons of* pylint errors in their tests that often don't make any sense (like `invalid-name`, ... see comment from before) and will most likely disable pylint for the file.

What I'm saying is that we should first try to go through *at least a few* test files and try to fix the pylint issues in there. If there are global `disable`s that need to be created in the process (and there most certainly will), they should be put in *before* we enable pylint on all `tests/` files. I'm not saying the fixes for *specific files* should be in *this PR*, we should simply do a case study of a few tests and find out what `disable`s are required.

cdce8p commented on tests/components/homekit/test_accessories.py:
```diff
@@ -2,6 +2,7 @@
 
 This includes tests for all mock object types.
 """
+# pylint: skip-file
```
This one

armills replied:
Oof. Good catch. My requirements were out of date. :flushed:

cdce8p commented on tests/components/homekit/test_homekit.py:
```diff
@@ -1,4 +1,5 @@
 """Tests for the HomeKit component."""
+# pylint: skip-file
```
Also this one

cdce8p reviewed (commented):
It seems like the comment was added to some files that don't throw any errors already. I just ran `pylint --msg-template='{path}' tests 2>/dev/null | grep '^tests' | uniq` locally and they wheren't in that list either.

cdce8p commented:
I agree with @OttoWinter. Before we start adding `# pylint: skip-file` to all error files, we should build the foundation first and stay away from `.tox` for the moment. It might help to create a seperate `pylintrc` for test runs.

emlove committed:
Disable pylint on our current broken test files

This change was made by the following bash script:

sed -i '0,/"""$/s//"""\n# pylint: skip-file/' $(pylint --msg-template='{path}' tests 2>/dev/null | grep '^tests' | uniq)

armills commented:
I don't disagree with any of that, but none of these ideas block turning on pylint for new PRs. This PR will never get merged if we want to bikeshed all the possible ways to handle the existing failures.

cdce8p commented:
That is not the point (at least I'm trying to make here). There are many pylint errors in our current test suite that aren't really errors for tests. Otto mentioned quiet a few here: https://github.com/home-assistant/home-assistant/pull/13093#discussion_r173664662

If those are caught probably we don't have to mark so many with `skip-file`.

OttoWinter commented:
> if we want to bikeshed all the possible ways to handle the existing failures.

I'm not saying all failures, just a case-study of a few (3-5) files. We would learn important stuff about what checks do and don't work (in that they produce lint errors that aren't really valid for tests, see comment) for the `tests/` folder. I've already gone through some files (in the comment i linked to before) and it wasn't too much work.

> I don't disagree with any of that, but none of these ideas block turning on pylint for new PRs.

I disagree with the last part of that statement there. As I said, this would just make new PRs a pain for newbies. Especially since we *can* go through some files without much work, I don't think we should make the PR process for others more work.

I'm copying the list of checks that are too verbose for `tests/` I found earlier here:

* `invalid-name` (too long test function names)
* `attribute-defined-outside-init`
* `protected-access` (obviously)
* `len-as-condition` (we shouldn't rely on truthy values)
* `redefined-outer-name` (fixtures)
* `no-member` (calling methods on `MagicMock`s)
* `no-value-for-parameter` (fixtures, again)
* probably lots more

armills commented:
I'm good if you want to push up some new commits to this branch to turn off those lint checks on `tests`.

OttoWinter commented:
I will take a look at that tomorrow. Need to catch some sleep (and learn) now, I have an exam coming up tomorrow 😅

Btw, looking back at my comments, they might be interpreted as being a bit "grumpy". That's not the case though, I'm just getting a bit tired 🛏💤

armills commented:
Hah, I know I've been terse as well. Let's review tomorrow once everyone's had some sleep. :smile:

balloob commented:
If we blacklist the ones that @OttoWinter listed, it might be ok. Because for example, I wouldn't want to change the test names or disable no-member for all the magic mocks etc.

armills commented:
Closing for the more comprehensive https://github.com/home-assistant/home-assistant/pull/13125 🙇
